enterprise/server/backends/pubsub/pubsub.go

```diff
@@ -138,8 +138,121 @@ func (p *ListPubSub) Subscribe(ctx context.Context, channelName string) interfac
 			}
 		}
 	}()
-	return &listSubscriber{
+	return &channelSubscriber{
 		cancel: cancel,
 		ch:     ch,
 	}
 }
+
+// To keep things simple for now, we maintain streams with a single value per element stored under the following key.
+const streamDataField = "data"
+
+type StreamPubSub struct {
+	rdb *redis.Client
+}
+
+// NewStreamPubSub creates a PubSub client based on a Redis-stream.
+func NewStreamPubSub(redisClient *redis.Client) *StreamPubSub {
+	return &StreamPubSub{rdb: redisClient}
+}
+
+func (p *StreamPubSub) subscribe(ctx context.Context, channelName string, startFromTail bool) interfaces.Subscriber {
+	ch := make(chan string)
+	ctx, cancel := context.WithCancel(ctx)
+
+	deliverMsg := func(msg *redis.XMessage) bool {
+		data, ok := msg.Values[streamDataField]
+		if !ok {
+			log.Errorf("Message %q on stream %q missing data field", msg.ID, channelName)
+			return false
+		}
+		str, ok := data.(string)
+		if !ok {
+			log.Errorf("Message %q on stream %q is of type %T, wanted string", msg.ID, channelName, data)
+			return false
+		}
+
+		select {
+		case ch <- str:
+			return true
+		case <-ctx.Done():
+			return false
+		}
+	}
+
+	go func() {
+		defer close(ch)
+
+		// Start from beginning of stream.
+		streamID := "0"
+
+		// If starting from tail, check if the stream has any elements.
+		// If it does then publish the last element and subscribe to messages following that element.
+		if startFromTail {
+			msgs, err := p.rdb.XRevRangeN(ctx, channelName, "+", "-", 1).Result()
+			if err != nil {
+				log.Errorf("Unable to retrieve last element of stream %q: %s", channelName, err)
+				return
+			}
+			if len(msgs) == 1 {
+				msg := msgs[0]
+				streamID = msg.ID
+				if ok := deliverMsg(&msg); !ok {
```

Comment:
can you shorten this to

if !deliverMsg(&msg) {
 return
}

Reply:
Done.